// flx-ui-datatable/esm5/lib/flx-ui-datatable.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ReactiveFormsModule, FormsModule } from '@angular/forms';
import { FlxUiDatatableComponent, ImageFallBack } from './flx-ui-datatable.component';
import { HttpModule } from '@angular/http';
import { FlxUiDatatableService } from './flx-ui-datatable.service';
import { Ceil } from './pipes/ceil.pipe';
import { HasLimitText } from './pipes/hasLimitText.pipe';
import { LimitText } from './pipes/limitText.pipe';
import { ProcessedData } from './pipes/processeddata.pipe';
import { PerfectScrollbarModule } from 'ngx-perfect-scrollbar';
import { PERFECT_SCROLLBAR_CONFIG } from 'ngx-perfect-scrollbar';
var /** @type {?} */ DEFAULT_PERFECT_SCROLLBAR_CONFIG = {
    suppressScrollX: true
};
var ɵ0 = DEFAULT_PERFECT_SCROLLBAR_CONFIG;
var FlxUiDatatableModule = /** @class */ (function () {
    function FlxUiDatatableModule() {
    }
    FlxUiDatatableModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule, ReactiveFormsModule, FormsModule, HttpModule, PerfectScrollbarModule
                    ],
                    declarations: [
                        FlxUiDatatableComponent, Ceil, ImageFallBack, HasLimitText, LimitText, ProcessedData
                    ],
                    exports: [FlxUiDatatableComponent, Ceil],
                    providers: [FlxUiDatatableService,
                        {
                            provide: PERFECT_SCROLLBAR_CONFIG,
                            useValue: ɵ0
                        }]
                },] },
    ];
    return FlxUiDatatableModule;
}());
export { FlxUiDatatableModule };
export { ɵ0 };
